codec-http/src/main/java/io/netty/handler/codec/http/HttpContentDecoder.java

```diff
@@ -145,7 +146,13 @@ protected void decode(ChannelHandlerContext ctx, HttpObject msg, List<Object> ou
             if (decoder == null) {
                 out.add(c.retain());
             } else {
-                decodeContent(c, out);
+                try {
+                    decodeContent(c, out);
+                } finally {
+                    if (!needRead) {
```

Comment:
I think this variable handling is inverted. `needRead` should start as `true`, and if `!out.isEmpty()` it should be set to `false`. Then in channelReadComplete() it should be set to `true` again.

If a read() causes three ByteBufs to be read (because using DefaultMaxBytesRecvByteBufAllocator), which calls decode() three times, the first two times may not put anything in `out` but the last decode does place something in `out`. It seems in that case we shouldn't call read.

But there's also the case where the _second_ bytebuf placed something in `out`. That seems like it shouldn't call read either.

Reply:
@ejona86 sure why not.